task_manager/task_manager.hh

```diff
@@ -0,0 +1,308 @@
+
+        status& get_status() noexcept {
+            return _impl->_status;
+        }
+
+        task_id get_parent_id() const noexcept {
+            if (_impl->_parent) {
+                return _impl->_parent->_impl->_status.id;
+            }
+            return task_id::create_null_id();
+        }
+
+        foreign_task_ptr& get_parent() noexcept {
+            return _impl->_parent;
+        }
+
+        void set_type(std::string type) noexcept {
+            _impl->_status.type = std::move(type);
+        }
+
+        void set_total_progress(double progress) {
+            _impl->_progress.total = progress;
+        }
+
+        void report_completion_progress(double progress) {
+            _impl->_progress.completed += progress;
+        }
+
+        void change_state(task_state state) noexcept {
+            _impl->_status.state = state;
+        }
+
+        void start() {
+            _impl->_status.start_time = db_clock::now();
+            _impl->_status.state = task_manager::task_state::running;
+
+            (void)with_gate(_impl->_module->async_gate(), [f = done(), module = _impl->_module, id = _impl->_status.id] () mutable {
+                return std::move(f).finally([module, id] {
+                    return sleep_abortable(module->get_task_manager().get_task_ttl(), module->get_task_manager().abort_source());
+                }).then_wrapped([module, id] (auto f) {
+                    f.ignore_ready_future();
+                    module->unregister_task(id);
+                });
+            });
+        }
+
+        static future<std::vector<task_id>> run_on_shards(task_ptr task, module_ptr module) {
+            return module->_tm.container().map([parent = seastar::make_foreign(task), module] (task_manager&) mutable -> foreign_task_ptr {
+                task_ptr child = parent->_impl->create_child(task_id::create_random_id(), std::move(parent)); // TODO Deexie the rest of fields
+                module->register_task(child);
+                child->start();
+                return seastar::make_foreign<task_ptr>(child);
+            }).then([task] (foreign_task_vector tasks) {
+                task->_impl->_children = std::move(tasks);
+                return make_ready_future<std::vector<task_id>>(boost::copy_range<std::vector<task_id>>(task->_impl->_children
+                    | boost::adaptors::transformed([] (foreign_task_ptr& child) { return child->_impl->_status.id; })));
+            });
+        }
+
+        module_ptr get_module() const noexcept {
+            return _impl->_module;
+        }
+
+        future<progress> get_progress() const {
+            return _impl->get_progress();
+        }
+        
+        is_abortable is_abortable() const noexcept {
+            return _impl->is_abortable();
+        };
+
+        future<> done() const noexcept {
+            return _impl->_done.get_shared_future();
+        }
+
+        void unregister() {
```

Comment:
Should be noexcept.
You don't want it to throw, especially if called in a destructor or on the atop() path

Reply:
Also, if we have task.unregister() why not have a register() interface
that will call the module register function?

(can use shared_from_this() if needed)